File: packages/core/src/coordinates/fake-coordinates.py
import random

def random_coord(x_range=(0, 500), y_range=(0, 500)):
    """Generate a random (x, y) coordinate within given ranges."""
    x = random.uniform(*x_range)
    y = random.uniform(*y_range)
    return round(x, 2), round(y, 2)


def random_svg_path(num_segments=1, seed=None, allow_close=True):
    """
    Generate a random SVG path string.
    Includes M (move), L (line), Q (quadratic), and C (cubic) segments.
    Each segment starts from the previous segment's end point.
    """
    if seed is not None:
        random.seed(seed)

    # Start point
    current = random_coord()
    d = [f"M{current[0]},{current[1]}"]

    for _ in range(num_segments):
        segment_type = random.choice(["L", "Q", "C"])
        segment_count = random.randint(1, 3)  # 1–3 consecutive of the same type

        for _ in range(segment_count):
            if segment_type == "L":
                # Line to
                x, y = random_coord()
                d.append(f"L{x},{y}")
                current = (x, y)

            elif segment_type == "Q":
                # Quadratic Bézier
                cx, cy = random_coord()
                x, y = random_coord()
                d.append(f"Q{cx},{cy} {x},{y}")
                current = (x, y)

            elif segment_type == "C":
                # Cubic Bézier
                c1x, c1y = random_coord()
                c2x, c2y = random_coord()
                x, y = random_coord()
                d.append(f"C{c1x},{c1y} {c2x},{c2y} {x},{y}")
                current = (x, y)

    if allow_close and random.random() < 0.3:
        d.append("Z")

    return " ".join(d)


def write_svg(paths, filename="random_paths.svg", size=(500, 500)):
    """Write a list of SVG path strings into an SVG file."""
    with open(filename, "w") as f:
        f.write(f'<svg xmlns="http://www.w3.org/2000/svg" width="{size[0]}" height="{size[1]}">\n')
        for d in paths:
            stroke = random.choice(["black", "red", "blue", "green"])
            f.write(f'  <path d="{d}" stroke="{stroke}" fill="none" stroke-width="2"/>\n')
        f.write("</svg>\n")


if __name__ == "__main__":
    # Generate multiple random paths
    paths = [random_svg_path(num_segments=random.randint(2, 5), seed=i) for i in range(200)]
    with open("fake-paths.txt", "w") as f:
        for path in paths:
            f.write(f"{path}\n")

File: packages/core/src/coordinates/fake-coordinates.ts
import { faker } from "@faker-js/faker";

type Coord = { x: number; y: number };

interface SvgPathOptions {
    numSegments?: number; // total groups of segments
    xRange?: [number, number];
    yRange?: [number, number];
    allowClose?: boolean;
    seed?: number;
}

/**
 * Generate a random (x, y) coordinate within given ranges.
 */
function randomCoord(
    xRange: [number, number],
    yRange: [number, number],
): Coord {
    const x = faker.number.float({
        min: xRange[0],
        max: xRange[1],
        multipleOf: 0.01,
    });
    const y = faker.number.float({
        min: yRange[0],
        max: yRange[1],
        multipleOf: 0.01,
    });
    return { x: Number(x.toFixed(2)), y: Number(y.toFixed(2)) };
}

/**
 * Generate a random SVG path.
 * Starts with a Move (M) and can contain Lines (L), Quadratic (Q), or Cubic (C) Bézier curves.
 */
export function generateRandomSvgPath(options: SvgPathOptions = {}): string {
    const {
        numSegments = 3,
        xRange = [0, 500],
        yRange = [0, 500],
        allowClose = true,
        seed,
    } = options;

    if (seed !== undefined) faker.seed(seed);

    let d: string[] = [];
    let current = randomCoord(xRange, yRange);
    d.push(`M${current.x},${current.y}`);

    for (let i = 0; i < numSegments; i++) {
        const segmentType = faker.helpers.arrayElement(["L", "Q", "C"]);
        const segmentCount = faker.number.int({ min: 1, max: 3 });

        for (let j = 0; j < segmentCount; j++) {
            if (segmentType === "L") {
                const p = randomCoord(xRange, yRange);
                d.push(`L${p.x},${p.y}`);
                current = p;
            } else if (segmentType === "Q") {
                const c = randomCoord(xRange, yRange);
                const p = randomCoord(xRange, yRange);
                d.push(`Q${c.x},${c.y} ${p.x},${p.y}`);
                current = p;
            } else if (segmentType === "C") {
                const c1 = randomCoord(xRange, yRange);
                const c2 = randomCoord(xRange, yRange);
                const p = randomCoord(xRange, yRange);
                d.push(`C${c1.x},${c1.y} ${c2.x},${c2.y} ${p.x},${p.y}`);
                current = p;
            }
        }
    }

    if (allowClose && faker.datatype.boolean()) {
        d.push("Z");
    }

    return d.join(" ");
}

/**
 * Example usage — generates multiple paths
 */
if (require.main === module) {
    for (let i = 0; i < 5; i++) {
        console.log(generateRandomSvgPath({ numSegments: 4, seed: i }));
    }
}
